1.1.1k fips jra disable algos with abort

jallisonciq opened the pull request:
Here is an updated version of the code that was reviewed yesterday by Solar and Maple.

The changes from that branch are the following:

1). For any FIPS unapproved algorithm that returns an error code on initialization, eg. 'int MD4_Init(), this code returns an error code on initialization and sets an internal error state which can be queried to determine if a FIPS unapproved algorithm was called.

2). Any attempt to use a FIPS unapproved algorithm that reported an error code on initialization will abort the program using:
```OpenSSLDie(__FILE__, __LINE__, "FATAL FIPS Unapproved algorithm called");```

eg. Calling any of MD4_Update() / MD4_Transform() / MD4_Final() after MD4_Init() returned a failure code will abort the calling program. Calling MD4_Update() / MD4_Transform() / MD4_Final() without calling MD4_Init() always aborts of course.

3). For any FIPS unapproved algorithm that does not return an error code on initialization. eg. Blowfish 'void BF_set_key()', or 'void IDEA_set_encrypt_key()' calling any initialization operation will immediately abort the calling program using OpenSSLDie(...as above..).

4). Any attempt to use a FIPS unapproved algorithm in any way that would transform data (i.e. any encrypt/decrypt/sign/verify/hash/message digest operations) will immediately abort the calling program using OpenSSLDie(...as above..). eg. 'void des_encrypt()' or 'void IDEA_cbc_encrypt()', or 'MD5_Update()'.

Note - this is for the underlying direct access algorithm functions. Normally, users of openssl create an instance of an algorithm by calling an instantiation function eg. EVP_md4(). All attempt to instantiate unapproved algorithms return NULL, thus disallowing access to the underlying algorithm via a returned openssl context pointer.

I believe this allows openssl to "fail safely" in the case of using FIPS unapproved algorithms, and not allow any calling application to believe it has successfully either instantiated an unapproved algorithm, or successfully used an unapproved algorithm to transform data in any way.

jallisonciq committed:
Prevent use of SHA1, MD2, MD4, MD5, ripemd160, sm3 in libcrypto when in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level DES functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add a new definition of BLOCK_CIPHER_def1, only enabled when FIPS_CIPHER_DISABLE is defined before including evp.h.

This allows us to only bear the cost of calling FIPS_mode()
and the associated locking when we know we the caller is
trying to instantiate a non-FIPS approved cipher.

Essentially all this change does is add a

+    if (FIPS_mode()) { \
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD); \
+        return NULL; \
+    } \

block into the generated definition of the
const EVP_CIPHER *EVP_##cname##_##mode(void) function,
but only if FIPS_CIPHER_DISABLE is defined before
inclusion of this header file.

We can now add:

 define FIPS_CIPHER_DISABLE 1

lines into all ciphers that instantiate themselves
as EVP as a line just before including include/crypto/evp.h,
and this will allow them to be disabled in FIPS mode.

Also adds the needed include lines to include/crypto/evp.h
in order for FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD)
to be correctly defined.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables ARIA cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables BF cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables CAST cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables DES cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables DES3 cipher in FIPS mode.

Also hand-disable EVP_des_ede, EVP_des_ede3 and EVP_des_ede3_wrap
separately.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables IDEA cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables RC2 cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables RC5 cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables SEED cipher in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add define FIPS_CIPHER_DISABLE 1 before including include "crypto/evp.h"

Disables SM4 cipher in FIPS mode.

Also disable EVP_sm4_ctr by hand.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable blake2b512 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable blake2s256 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable md5_sha1 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable mdc2 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable whirlpool in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable camellia in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable chacha20 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable "old" functions for EVP_bf_cfb, EVP_des_cfb, EVP_des_ede3_cfb, EVP_des_ede_cfb, EVP_des_ede_cfb, EVP_idea_cfb, EVP_rc2_cfb, EVP_cast5_cfb in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable rc4 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable RC4_HMAC_MD5 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable XCBC in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low level aria functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level BF (blowfish) functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
EVP_add_digest() shouldn't crash when given a NULL pointer.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level BLAKE2 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level CAMELLIA code in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level idea functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level cast functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable ChaCha20_ctr32() in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level mdc2 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level functions for Poly1305 in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level RC2 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level RC4 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level RC5 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level SEED algorithm functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level SM2 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Prevent use of sm3_block_data_order() in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable use of low-level sm4 functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level WHIRLPOOL functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable scrypt_pkey_method in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable dsa_pkey_method in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable ed25519 and ec*488 pkey methods in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level SIPHASH functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable siphash_pkey_method in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Create a new standard_methods_fips array, for EVP_PKEY methods. Only used in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Add new macro BLOCK_CIPHER_custom_fips() used to return NULL when trying to instantiate AES ocb mode when FIPS is enabled.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Disable low-level AES_ige_encrypt/AES_ige_decrypt functions in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Cleanup speed output - don't show disabled algorithms in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Use a separate array "standard_methods_fips" for getting ASN1 key methods when in FIPS mode.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

jallisonciq committed:
Warning messages cleanup - prod version.

Signed-off-by: Jeremy Allison <jallison@ciq.com>

PlaidCat commented:
Original PR that prompted these changes:
https://github.com/ciq-rocky-fips/openssl/pull/4

Rest of discussion was CIQ internal Slack conversation.

PlaidCat commented on apps/speed.c:
I'm not super thrilled with the below break code but I'm not sure there is better way given some of the surrounding code in a couple of the codes
```
for ()
  if fips()
     break;
```

jallisonciq replied:
Yep. Seemed cleaner than a "goto out".

PlaidCat commented on crypto/aria/aria.c:
All good only exposed functions in header are covered
https://github.com/ciq-rocky-fips/openssl/blob/d25d84a8762f11fa1c0aaff80bb0fa466bb25264/include/crypto/aria.h

PlaidCat commented on crypto/asn1/standard_methods.h:
```diff
@@ -59,3 +59,61 @@ static const EVP_PKEY_ASN1_METHOD *standard_methods[] = {
 #endif
 };
 
+static const EVP_PKEY_ASN1_METHOD *standard_methods_fips[] = {
```
is this necessary to break out into its own?
We don't plan on releasing this outside of FIPS right and even if we did we would not compile with the FIPS options so they would still work out right ?

If the is all true the we could eliminate the changes to `crypto/asn1/ameth_lib.c**crypto/asn1/ameth_lib.c`

If there is an additional detail I'm missing that would be helpful, thanks

jallisonciq replied:
The reason I added it to crypto/asn1/standard_methods.h instead of adding it to the .c file was that it is *much* easier to confirm that the array added here is identical to the "standard_methods" array with the addition of the #ifdefs around FIPS to remove the unwanted algos.

If it was in the .c file, then it would be much easier for a subsequent CVE patch to maybe change the "standard_methods" array in the .h file, and for us to forget to make the same change in the "standard_methods_fips array" at the same time.

Keeping the two arrays together makes it easier to keep them in sync. This is just my experience of many years of code changes like this :-).

PlaidCat replied:
I see you point, we'd just need to make sure that if we're addressing CVEs and it modifies the array[] `static const EVP_PKEY_ASN1_METHOD *standard_methods` that we also replicate the changes to `static const EVP_PKEY_ASN1_METHOD *standard_methods_fips` 

I think maybe that would have been a case to integrate the `#ifndef FIPS` into the `static const EVP_PKEY_ASN1_METHOD *standard_methods` it would conflict and force us to make sure we know how to integrate the changes. 
Then we could also remove the `if else` in the `.c`.

If this is a method you're comfortable with maintaining then I will not make a fuss.

jallisonciq replied:
I just realized I didn't answer the first question: "is this necessary to break out into its own?"

Yes it has to be a separate array (with the changes to crypto/asn1/ameth_lib.c**crypto/asn1/ameth_lib.c also). The reason is that non-FIPS mode still has to work with this library, as this is how all the tests are run in the rpm build.

I originally tried to use only a truncated array, #ifdef'ed out, and this caused all the tests in the rpm build to fail.

So we need the original array for non-fips mode (it still has to work in that mode), and the truncated array for when we switch into fips mode.

PlaidCat replied:
Cool thanks for the clarification here

PlaidCat commented on crypto/blake2/blake2b.c:
```diff
@@ -81,6 +89,11 @@ static void blake2b_init_param(BLAKE2B_CTX *S, const BLAKE2B_PARAM *P)
 /* Initialize the hashing context.  Always returns 1. */
```
I hate this comment.

We're now returning a zero but it seems that the return is expected a qualified structure that may be just randomly allocated space and allow the other functions like `blake2b_compress` to run unexpectedly with random input.

See the path from `blake2b_init_param` to `blake2_init0` that is done prior to the `return 1`

We might need to add the fips check to `static void black2b_compress`
```
    if (FIPS_mode()) {
        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
        return 0;
    }
```

jallisonciq replied:
Well it's not my comment :-). 

Also, the definition of blake2b_compress() is:

static void blake2b_compress()

which means there's no way to get to this externally (it's defined static). So I don't think the fips check is needed in there.

Also, blake2b_init_param() is static as well.

PlaidCat replied:
Yeah its called at the end of this `init` function.

I missed the static part of the compress when I saw that the fips check was missing and the weird statement in the comment that the `inti` will always return 1 then saw that `BLAKE2B_CTX *c` was memset with in `int BLAKE2b_Init(BLAKE2B_CTX *c)` that I then forgot the start of the fuction and jumped into how it could have been used in that function.

PlaidCat replied:
Gonna resolve

PlaidCat commented on crypto/blake2/blake2s.c:
```diff
@@ -79,6 +87,11 @@ int BLAKE2s_Init(BLAKE2S_CTX *c)
 {
```
Same issues as in `blake2b` 

I hate this comment.

We're now returning a zero but it seems that the return is expected a qualified structure that may be just randomly allocated space and allow the other functions like `blake2b_compress` to run unexpectedly with random input.

See the path from `blake2b_init_param` to `blake2_init0` that is done prior to the `return 1`

We might need to add the fips check to `static void black2b_compress`
```
    if (FIPS_mode()) {
        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
        return 0;
    }
```

jallisonciq replied:
Again. Static functions. "You can't get there from here". All entry points are covered with OpenSSLdie or error returns already.

PlaidCat replied:
Yeah its called at the end of this `init` function.

I missed the static part of the compress when I saw that the fips check was missing and the weird statement in the comment that the `inti` will always return 1 then saw that `BLAKE2B_CTX *c` was memset with in `int BLAKE2b_Init(BLAKE2B_CTX *c)` that I then forgot the start of the fuction and jumped into how it could have been used in that function.

resolving

jason-rodri commented on crypto/bf/bf_enc.c:
```diff
@@ -26,6 +26,12 @@ void BF_encrypt(BF_LONG *data, const BF_KEY *key)
+    if (FIPS_mode()) {
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
+        OpenSSLDie(__FILE__, __LINE__, "FATAL FIPS Unapproved algorithm called");
+        return;
```
Do we really need the return?

jallisonciq replied:
Hmmm. I always use return in void cases, even after the call to OpenSSLDie() as a "belt and braces" approach. I just don't want it blundering into subsequent code for *any* reason. Judgement call - I'll go with whatever you desire on this.

PlaidCat replied:
I think it makes it clear that we have no intention to execute the reset of the function and if someone didn't know exactly what `OpenSSLDie` does (yes its self explanatory) even if that `return` is never executed.

jason-rodri commented on crypto/cast/cast_local.h:
```diff
@@ -7,6 +7,13 @@
 
+# include "openssl/opensslconf.h"
+
+#ifdef OPENSSL_FIPS
```
Is this extra code? I don't see anything in this file that would require fips.h/err.h

jallisonciq replied:
Nope. Removing it gives:
```
cast/c_cfb64.d.tmp -MT crypto/cast/c_cfb64.o -c -o crypto/cast/c_cfb64.o crypto/cast/c_cfb64.c
crypto/cast/c_cfb64.c: In function 'CAST_cfb64_encrypt':
crypto/cast/c_cfb64.c:29:9: warning: implicit declaration of function 'FIPS_mode' [-Wimplicit-function-declaration]
   29 |     if (FIPS_mode()) {
      |         ^~~~~~~~~
crypto/cast/c_cfb64.c:30:9: warning: implicit declaration of function 'FIPSerr' [-Wimplicit-function-declaration]
   30 |         FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
etc....
```

It removes warnings (I like a clean compile :-).

PlaidCat replied:
this is a local_include for most of the `crypto/cast/*.c`

solardiz commented on crypto/evp/pmeth_lib.c:
```diff
@@ -70,6 +70,69 @@ static pmeth_fn standard_methods[] = {
+#endif
+#ifndef OPENSSL_NO_SCRYPT
+#ifndef OPENSSL_FIPS
+    scrypt_pkey_method,
```
FWIW, scrypt uses FIPS-approved SHA-256 on the outer layer and its cryptographic security for the password input is guaranteed by SHA-256, HMAC, PBKDF2. I don't know if this possibly makes it FIPS-compliant despite of it also using Salsa20 internally?

I am using a similar argument to justify usage of yescrypt where NIST-approved primitives are needed - in there, we have entropy bypasses for both password and salt inputs, so cryptographic security is provably guaranteed by SHA-256, HMAC, PBKDF2, whereas the actually most important (against GPUs, etc.) internal processing is "non-cryptographic".

I think OpenSSL-provided scrypt isn't used much yet (it's a recent addition to OpenSSL), so I am more interested in whether this is a stance we'd like to defend in general. We're going to have yescrypt for user passwords on EL 10 by default.

jallisonciq replied:
It's not FIPS compliant. It's not listed in our security policy, so we have to disable it.

solardiz commented on crypto/kdf/scrypt.c:
```diff
@@ -506,6 +514,10 @@ static int scrypt_alg(const char *pass, size_t passlen,
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
+        return NULL;
+    }
     return &scrypt_pkey_meth;
```
We might want to leave scrypt allowed. See my comment above.

jallisonciq replied:
No, it's not listed in our security policy so we must disable it. Jason can follow-up here if needed.

solardiz commented on crypto/md2/md2_dgst.c:
```diff
@@ -70,6 +77,10 @@ const char *MD2_options(void)
 {
+    if (FIPS_mode()) {
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
+        return 0;
```
In `MD*_Init()` and the like, can we perhaps allow the context initialization to complete, and then return error? Would be very slightly safer in terms of not leaving uninitialized data that an application may use somehow. I hope it's not a FIPS violation merely to initialize a struct with constants or zeroize it?

jallisonciq replied:
I don't think that's the right approach. Functions that return failure should *never* modify incoming data. In Samba, this is a golden rule. We only copy initialized data into "out" parameters just before returning success.

On failure return, no change should be made to "out" parameters. I think you'll find this is the rule in glibc and most other libraries too.

jallisonciq replied:
Resolved with the follow-up compromise patch.

solardiz commented on crypto/md4/md4_dgst.c:
```diff
@@ -22,6 +30,10 @@
 int MD4_Init(MD4_CTX *c)
 {
+    if (FIPS_mode()) {
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
```
For MD4 and MD5, I don't see the corresponding pieces in Update and Final that we have for MD2 here - at least not in this PR's context. Are we missing them, or did we already have them before this PR?

jallisonciq replied:
The Update, Transform and Final methods are constructed via inline functions in crypto/md32_common.h.

We update the HASH_UPDATE, HASH_TRANSFORM and HASH_FINAL definitions in  crypto/md32_common.h to add the needed code.

jallisonciq replied:
git show ea3f8d0c1556f908f8356b28c7dd22b3f300adcc "Prevent use of SHA1, MD2, MD4, MD5, ripemd160, sm3 in libcrypto when in FIPS mode." is the commit that does this.

solardiz commented on crypto/poly1305/poly1305_base2_44.c:
```diff
@@ -59,6 +66,11 @@ int poly1305_init(void *ctx, const unsigned char key[16])
 
+    if (FIPS_mode()) {
+        FIPSerr(ERR_LIB_FIPS, FIPS_R_NON_FIPS_METHOD);
+        return 1;
```
Is there a specific convention to return 1 on error here, or are we inventing it just now seeing how 0 is returned on success? In the latter case, it may be slightly safer to return -1.

jallisonciq replied:
The other poly1305_XXX functions return 1 on error. Principle of least surprises applies here. Arbitrarily returning -1 is unexpected IMHO.

solardiz commented on crypto/rsa/rsa_oaep.c:
```diff
@@ -129,6 +121,15 @@ int ossl_rsa_padding_add_PKCS1_OAEP_mgf1_ex(unsigned char *to, int tlen,
+{
+    return ossl_rsa_padding_add_PKCS1_OAEP_mgf1_ex(to, tlen, from, flen,
+                                                   param, plen, NULL, NULL);
+}
```
Why are we moving `RSA_padding_add_PKCS1_OAEP` down in this source file? I see we're moving it to after the function it calls, but I guess this somehow compiled before anyway, perhaps there's a prototype somewhere?

jallisonciq replied:
Yes, this is a warning removal. Without these there are so many warnings it become hard to see real bugs.

jallisonciq committed:
For initialization functions that return an error code but take a passed in context pointer, null out the context memory before returning an error.

This (IMHO) goes against good programming practice, as
error code return paths should not modify in/out parameters.
So I'm doing this under protest :-).

Affected functions are: BLAKE2b_Init(), BLAKE2s_Init(), MD2_Init(),
MD4_Init(), MD5_Init(), MDC2_Init(), RIPEMD160_Init(), SHA1_Init(),
SipHash_Init(), sm3_init(), WHIRLPOOL_Init().

Signed-off-by: Jeremy Allison <jallison@ciq.com>

PlaidCat reviewed (approved):
That took a while but I like the updates thanks for the changes.
